apps/sms/services/activity/service.js

```diff
@@ -0,0 +1,85 @@
+/* global threads */
+
+(function(exports) {
+  'use strict';
+
+  /**
+   * Name of the service that is responsible for managing activities.
+   * @type {string}
+   */
+  const SERVICE_NAME = 'activity-service';
+
+  /**
+   * Definition of all available private members.
+   * @type {Object}
+   */
+  const priv = Object.freeze({
+    service: Symbol('ActivityService.service'),
+    activityRequest: Symbol('ActivityService.activityRequest'),
+
+    onPostResult: Symbol('ActivityService.onPostResult'),
+    onPostError: Symbol('ActivityService.onPostError'),
+    onActivityRequest: Symbol('ActivityService.onActivityRequest')
+  });
+
+  var ActivityService = function() {
+    var service = this[priv.service] = threads.service(SERVICE_NAME);
+
+    service.method('postResult', this[priv.onPostResult].bind(this));
+    service.method('postError', this[priv.onPostError].bind(this));
+
+    exports.navigator.mozSetMessageHandler(
+      'activity', this[priv.onActivityRequest].bind(this)
+    );
+  };
+
+  /**
+   * Posts activity request result.
+   * @param {*} result Data to post as activity request result.
+   * @private
+   */
+  ActivityService.prototype[priv.onPostResult] = function(result) {
+    var activityRequest = this[priv.activityRequest];
+
+    if (!activityRequest) {
+      throw new Error('Application is not in activity mode!');
+    }
+
+    activityRequest.postResult(result || { success: true });
+
+    this[priv.activityRequest] = null;
+  };
+
+  /**
+   * Posts activity request error.
+   * @param {*} error Error to post as activity request error.
+   * @private
+   */
+  ActivityService.prototype[priv.onPostError] = function(error) {
+    var activityRequest = this[priv.activityRequest];
+
+    if (!activityRequest) {
+      throw new Error('Application is not in activity mode!');
+    }
+
+    activityRequest.postError(error);
+
+    this[priv.activityRequest] = null;
+  };
+
+  /**
+   * Handler that fires once app receives activity request via system message.
+   * @param {ActivityRequestHandler} request Activity request.
+   * @private
+   */
+  ActivityService.prototype[priv.onActivityRequest] = function(request) {
+    this[priv.activityRequest] = request;
+
+    // Talk to Wilson to understand how to fix this.
+    setTimeout(() => {
```

Comment:
Otherwise message isn't dispatched, likely because of https://github.com/gaia-components/threads/blob/4a440ddf3b448ce235f709bae06440a787076447/threads.js#L1749